### ISSUE TITLE
Show warnings when doing an action or playing a card that may not be "useful"

### ISSUE THREAD

**Issue by Lynesth** [bafolts/terraforming-mars#1642]
## The idea
That title is probably a bit vague, so let me try to explain better.
The idea is to show a warning if the card/action you are playing is probably not going to do what you expect. Here are some examples of such moves:

- Using the standard project **Asteroid** when temperature is already maxed.
- Activating **Martian Rails** without a city on mars.
- Playing **Energy Tapping** when none of your opponents have energy production.
- Playing **Virus** against a opponent that has **Protected Habitats**.

Not saying that there's no value at all in playing those, but it's most likely not going to do what you expected if you're not focused enough.

## What exactly
Here's the proposed list of things to track and show warnings for:
- Global Parameter (O2, Temp, Venus, Ocean) already maxed (or not getting full effect from card)
- Removing plants/microbes/animals while your only valid target has Protected Habitats (or it's animals on Pets)
- Removing resources without any valid target, like playing Virus but noone has microbes/plants (not counting yourself)
- Decreasing anyone's production where the only valid target is yourself
- Gaining resources/production for every city without any city
- Gaining resources/production/bonus for every colony without any colony
- Gaining resources/production for each specific tag without said tag (or without enough tags to get at least 1)
- On cards that place Oceans + Remove plants, if one of the available targets has Arctic Algae
- Adding resources while having no cards able to receive them
- Look at top card to buy it without enough money to buy the card

I think we should only track what the card/action does and not subsequent effects that derives from the play. Example: If a card places a Greenery, we should **not** check if oxygen is already maxed.

## How (technically)
There's currently two proposals that need inputs and opinions:

#### Card by card
Add a `checkWarnings: (Game, Player) => void` method to cards that will populate a `warningText?: string | undefined` property.
Call `checkWarning` after checking if the card/action is playable. Use `warningText` to print the warning message.

#### A class to check them all
Create a new class that will be called to check for warnings, using `checkWarnings: (Game, Player, ICard) => string | undefined`.
Add a few properties to cards (non exhaustive list): `addsOxygen?: number`, `addsTemperature?: number`, `addsOcean?: number`, `addsVenus?: number`, `removesPlants?: number`, etc. Use string returned`checkWarnings to print the warning message.

## How (visually)
The proposal is to have a message that doesn't is simple, visible and doesn't require you to take any additional action.
It would just be a small warning above the button that validates the action you want to take.
Optimally, it would be turned ON with an options in game preferences, and that option would be **OFF by default**.

Here's a rough draft of this in play:
![firefox_2020-10-17_21-59-36](https://user-images.githubusercontent.com/5318258/96335447-62abe900-10c4-11eb-8843-72594471dcd8.png)

## TL;DR
Read it if you want your opinion to matter!

**Comment by vsrisuknimit** [comment]
This seems like a great idea. Red text on black should be improved. If there is a way to change the color of the "Play card" button, that would be nice too.

**Comment by kevinb9n** [comment]
This comment not intended to distract anyone from doing the easiest thing for the clearest cases. Just to maybe drum up more interest in the long-term potential, or something.

To fill out the laundry list (apologies for perhaps unimportant examples, just because this is fun):

* Ant/Predator has to eat its own
* Inventors Guild / Business Network without the $ you need (bit me as Polyphemos recently)
* The powered-up Nitrogen-Rich Asteroid and Io Sulphur Research stats (bit a friend recently)
* Attack component of Flooding when no eligible player has money
* ... of Comet for Venus when no other player has both Venus tag and any money
* Indentured Workers / Conscription when no other hand card is playable
* Special Design when no card in hand is 1 or 2 steps out of range
* Productive Outpost with no colonies
* Event Analysts when all parameters are maxed
* Ecology Experts prelude when no card in hand has min global reqts

Most of these situations seem to generalize as: "this icon on this card won't currently have {any,full} effect." Just to make this harder :-) ponder the UI just displaying a "red X" or the equivalent over such icons. e.g. if there are 8 oceans on the board then you'd see

![Two ocean icons, only one crossed out](https://user-images.githubusercontent.com/934551/96339504-086e4200-104a-11eb-8c04-37a1fe32526c.png)

For the greenery/o2 case you suggest NOT doing, well, at least there would be a way to depict it if we wanted to

![Greenery icon with only o2 crossed out](https://user-images.githubusercontent.com/934551/96342265-479e9200-104e-11eb-9ee0-5e237ecb3adf.png)

This could be nice even for other cases *outside* of the actual topic of this thread, when a "warning" isn't exactly appropriate, but it can just show that your options are currently more limited than you might have thought.

![Imported Hydrogen with only microbe icon crossed out](https://user-images.githubusercontent.com/934551/96339615-b7ab1900-104a-11eb-92e0-b1ed4e4fb5ef.png)

Of course, this would be a lot work to do... but less text to have to be translated...

**Comment by kberg** [comment]
This was discussed on the dev forum, I'll repeat my major thoughts that differ from the full proposal.

## Slippery slope?

> I'm trying to think about whether that's a slippery slope. I don't see it yet, but I'm just letting my nervous instinct drive my thinking.

It's something to look out for. Peoples' preferences for what's a proper recommendation or not -- and I think part of that is that this design proposes warning about "useful" operations. This proposal should be clear that it's focused on things like "No, you can't raise the temperature anymore, why are you using Deimos Down?" to "You're raising your heat 7 steps after the entire planet has been terraformed, and nobody's going for the Thermalist award, anyway, so use your 35 MC better." So, please use a more clear term than "useful" so we can be in agreement.

## We can do better? Can we do better?

> Think about this: an action that defines raising the temperature one step.
> With one bit of data on the card we can:
> 1. Execute the behavior without having to write the code
> 2. Provide a visual cue to the caller about max temp
> 3. Provide iconography for the Card describing the action
> 4. Provide text for the card describing the action.
> Now -- THAT seems powerful.

@chosta is doing a lot of work already reducing duplication around cards, and my view is this proposal is half-way to a much more powerful one. It comes with its own challenges, but I say go back and think bigger. Don't write custom `checkWarning` calls. Let it be driven by metadata.

**Comment by kevinb9n** [comment]
> > I'm trying to think about whether that's a slippery slope. I don't see it yet, but I'm just letting my nervous instinct drive my thinking.

I think this feature can be limited to only situations that fit the statement
"This icon (element) of this action will (near-)provably have no effect."

True, we already have 20 bullet points above that do all seem to fit this (I think?), and that already seems like a lot. But they all seem clearly useful to me. If there's another 15 we're missing that *still fit the bill* then we'd probably want them too.

What we can avoid doing is branching out into warning about *partial* effects. Like the only player you can Flood has only 3 MC, not the full 4!  (Gasp!). That would be going too far.

**Comment by Lynesth** [comment]
@kevinb9n Thank you for your input.
I'm really not in favour of modifying cards iconography so the red cross over icons is not a good idea to me.
Some of your ideas were already included in the list but I added a few others. Although:
- Prelude cards have to be played anyway so there's no point in checking them.
- I'm not sure that Influence increase (eg. **Event Analysts**) should be checked as you would still get it and the last generation is pretty obvious to everyone.
- I'm also not sure about effects that modify the next card you play (eg. **Indentured Workers** or **Special Design**).

------

@kberg Thank you for participating on this matter !
> This was discussed on the dev forum

We were told to make it a Github issue, so there it is.
> Slippery slope?
> This proposal should be clear 
> So, please use a more clear term than "useful" so we can be in agreement.

I literally started the issue with _`That title is probably a bit vague, so let me try to explain better`_.
I thought I made it pretty clear by giving a list of things that should be checked for. And those are direct effects of the card, as stated: _`we should only track what the card/action does and not subsequent effects that derives from the play`_.
So don't you think your **Soletta** example doesn't fit the criterias and might be a bit far-fetched ?

> @chosta is doing a lot of work already reducing duplication around cards
> Don't write custom checkWarning calls. Let it be driven by metadata.

That joins the 2nd suggestion of making checks driven by data on cards, which I think would be a good way of doing it.
Maybe `canPlay` and `canAct` could be generalized as well somehow at some point.

**Comment by kberg** [comment]
> @kberg Thank you for participating on this matter !

You're welcome!

> > This was discussed on the dev forum
> 
> We were told to make it a Github issue, so there it is.

No problem -- I was just making that clear to other people.

> > Slippery slope?
> > This proposal should be clear
> > So, please use a more clear term than "useful" so we can be in agreement.
> 
> I literally started the issue with _`That title is probably a bit vague, so let me try to explain better`_.
> I thought I made it pretty clear by giving a list of things that should be checked for. And those are direct effects of the card, as stated: _`we should only track what the card/action does and not subsequent effects that derives from the play`_.
> So don't you think your **Soletta** example doesn't fit the criterias and might be a bit far-fetched ?

Oh, certainly! I'm asking that criteria be more clearly laid out so we can get from _vague_ to _shared understanding!_

**Comment by chosta** [comment]
@Lynesth should we close this one or you expect the discussion to continue :) ?
